File: packages/java/src/main/java/fyi/sererr/CapturedError.java
package fyi.sererr;

import java.util.List;
import java.util.Objects;
import java.util.Optional;

/**
 * A captured error.
 *
 * <p>Field-compatible with {@code sererr.v1.CapturedError}. A cause
 * chain is represented as a {@code List<CapturedError>}, most-causal-
 * first — the originating caught error is the LAST element. Chain
 * position is identified by {@link ExceptionMechanism#exceptionId()} /
 * {@link ExceptionMechanism#parentId()}.
 */
public record CapturedError(
        String type,
        String message,
        List<StackFrame> frames,
        Optional<ExceptionMechanism> mechanism,
        String release,
        String serverName
) {
    public CapturedError {
        Objects.requireNonNull(type, "type");
        Objects.requireNonNull(message, "message");
        Objects.requireNonNull(frames, "frames");
        Objects.requireNonNull(mechanism, "mechanism");
        Objects.requireNonNull(release, "release");
        Objects.requireNonNull(serverName, "serverName");
        frames = List.copyOf(frames);
    }

    /** All-defaults captured error (all zero-value fields, no mechanism). */
    public static CapturedError empty() {
        return new CapturedError("", "", List.of(), Optional.empty(), "", "");
    }

    public static Builder builder() {
        return new Builder();
    }

    public static final class Builder {
        private String type = "";
        private String message = "";
        private List<StackFrame> frames = List.of();
        private Optional<ExceptionMechanism> mechanism = Optional.empty();
        private String release = "";
        private String serverName = "";

        public Builder type(String v) { this.type = v; return this; }
        public Builder message(String v) { this.message = v; return this; }
        public Builder frames(List<StackFrame> v) { this.frames = v; return this; }
        public Builder mechanism(ExceptionMechanism v) { this.mechanism = Optional.ofNullable(v); return this; }
        public Builder mechanism(Optional<ExceptionMechanism> v) { this.mechanism = v; return this; }
        public Builder release(String v) { this.release = v; return this; }
        public Builder serverName(String v) { this.serverName = v; return this; }

        public CapturedError build() {
            return new CapturedError(type, message, frames, mechanism, release, serverName);
        }
    }
}

File: packages/csharp/src/CapturedError.cs
namespace Sererr;

/// <summary>
/// A captured error: type, message, frames, mechanism, plus inlined
/// event-level metadata (release / server_name).
///
/// Cause chains are a flat <see cref="IReadOnlyList{T}"/> of
/// <see cref="CapturedError"/> on the enclosing collection
/// (most-causal-first; the originating caught error is the LAST
/// element — matches Sentry's <c>exception.values</c> convention).
/// </summary>
public sealed record CapturedError
{
    /// <summary>Error class/type name.</summary>
    public string Type { get; init; } = "";

    /// <summary>Short rendering of the error itself, NOT including the chain.</summary>
    public string Message { get; init; } = "";

    /// <summary>Frames, most-recent-call-first.</summary>
    public IReadOnlyList<StackFrame> Frames { get; init; } = Array.Empty<StackFrame>();

    /// <summary>Mechanism describing how the exception was captured.</summary>
    public ExceptionMechanism? Mechanism { get; init; }

    /// <summary>Build identifier (e.g. semver or git SHA).</summary>
    public string Release { get; init; } = "";

    /// <summary>Producing host / pod name.</summary>
    public string ServerName { get; init; } = "";

    public bool Equals(CapturedError? other)
    {
        if (ReferenceEquals(this, other)) return true;
        if (other is null) return false;
        return Type == other.Type
            && Message == other.Message
            && Frames.SequenceEqual(other.Frames)
            && Equals(Mechanism, other.Mechanism)
            && Release == other.Release
            && ServerName == other.ServerName;
    }

    public override int GetHashCode()
    {
        var hc = new HashCode();
        hc.Add(Type);
        hc.Add(Message);
        foreach (var f in Frames) hc.Add(f);
        hc.Add(Mechanism);
        hc.Add(Release);
        hc.Add(ServerName);
        return hc.ToHashCode();
    }
}
